rfd/0218-access-list-members-iac.md

```diff
@@ -0,0 +1,249 @@
+state: draft
+---
+
+# RFD 218 - Access List members IaC
+
+## Required Approvals
+
+* Engineering: @r0mant && @smallinsky
+* Product: @r0mant
+
+## What
+
+Ability to manage Access List members with Terraform. This is enterprise-only feature.
+
+## Why
+
+Currently the Access List membership model is very dynamic in nature. Periodic membership reviews
+are required, membership can expire, and custom dynamic eligibility criteria can be specified.
+Because of that, the IaC approach to Access List membership was not provided so far and we don't
+have a good way to introduce it for the Access Lists in their current form.
+
+Manual management of Access List membership doesn't always scale. There are ways of proper
+structuring teams as Access Lists and them using the nested Access List concept to assign teams to
+resources, but that doesn't work when users are managed externally.
+
+The concept of dynamically assigning users to Access Lists (something like membership criteria)
+also won't scale when users are managed externally in large organizations. For example Microsoft
+Entra ID won't display any groups in SAML assertion [if the user is assigned to more than 150
+groups](https://learn.microsoft.com/en-us/entra/identity/hybrid/connect/how-to-connect-fed-group-claims).
+
+A new concept, a ***static Access Lists***, is introduced to overcome the outlined limitations. The
+idea is to have Access List with a *sub_kind* set to "static". Creating the Access List with the
+new *sub_kind* will disable the dynamic features of the *static* Access List (like reviews,
+expiration, or eligibility criteria). This will make it possible to manage such Access Lists using
+IaC tools. 
+
+## Details
+
+### Glossary
+
+- ***static Access List*** - Access List with the *.spec.type* field set to "static".
+
+### UX
+
+#### Terraform
+
+There will be a new Terraform resource named `teleport_access_list_member`. 
+
+```hcl
+resource "teleport_access_list" "crane_operation" {
+  header = {
+    version = "v1"
+    metadata = {
+      name = "crane-operation"
+    }
+  }
+  spec = {
+    # type must be set to "static" to manage members with Terraform.
+    type = "static"
+    title = "Crane operation"
+    description = "Used to grant access to the crane."
+    grants = {
+      roles = ["crane-operator"]
+    }
+    # membership_requires is optional.
+    membership_requires = {
+      roles = ["crane-operation-license"]
+    }
+  }
+}
+
+resource "teleport_access_list_member" "crane_operator" {
+  header = {
+    version = "v1"
+    metadata = {
+      name = "crane-operator" 
+    }
+  }
+  spec = {
+    access_list = teleport_access_list.crane_operation.id
+    // membership_kind is 1 for "MEMBERSHIP_KIND_USER" or 2 for "MEMBERSHIP_KIND_LIST"
+    membership_kind = 1
+    // expires is optional. The member will stay in the list after it expires but will lose the
+    // grants. expires can be updated.
+    expires = "2025-07-28T22:00:00Z"
+  }
+}
```

Comment:
Can we include an example of how a user would define an access list with a member that is an access list created by an integration (e.g. synced from Okta or Entra)?

Reply:
Example added in https://github.com/gravitational/teleport/pull/55868